### PR TITLE
Allows Organ Harvesters to Ignore Clothing When Emagged + Harvest Head Last

### LINKED ISSUE #43423: Emagged Organ Harvesters Still Require Their Victims to Have No Items
https://github.com/tgstation/tgstation/blob/ccadcf61d3d383191d950a8ac5a350b0213ba9e9/code/game/machinery/harvester.dm

There's an "var/allow_clothing" in the code for organ harvesters, but it's set to FALSE by default and can't be changed to TRUE without admin intervention.

Also, I think the victim's head comes off before the organs are extracted from the chest of the victim. Please change the ordering of that; I want to listen to my victims' screams as a machine rips their heart out of their chest before their very eyes, but they can't really scream without a head.

### PR DESCRIPTION
**Pull request by bobbah**
## About The Pull Request
Fixes #43423.

The organ harvester will now harvest bodies that have items on them so long as it is emagged. It will also now harvest the head last, when applicable, in most cases.

## Why It's Good For The Game
Ability to harvest with clothing on seems to be reasonable for an emagged harvester. Harvesting the head last is merely a flavor change, but a welcome one.

## Changelog
:cl: bobbahbrown
add: Organ harvesters are now rumored to ignore abiotic items on bodies, living or dead, if the safety sensors are broken.
tweak: Organ harvesters now harvest the head last.
tweak: Organ harvesters now require bodies to be dragged onto them, with a delay similar to gibbers and disposal bins.
/:cl:

### CODE REVIEW

**Comment by ATH1909** [comment]
Huzzah!

**Comment by Time-Green** [comment]
You can instantly sentence soneone to death by drag & dropping them on an emagged organ harvester, and no one would be any wiser. 

The only counter would be jumping out very fast before it starts up, crowbarring it open or screaming over the radio and pray someone clones you.

But I kinda wanna see it happen though

**Comment by anconfuzedrock** [comment]
Imagine reaching a state of such unbalance that nobody even considers the negative consequences of turning a sleeper into an instakill machine

**Comment by bobbahbrown** [comment]
> You can instantly sentence soneone to death by drag & dropping them on an emagged organ harvester, and no one would be any wiser.
> 
> The only counter would be jumping out very fast before it starts up, crowbarring it open or screaming over the radio and pray someone clones you.
> 
> But I kinda wanna see it happen though



> Imagine reaching a state of such unbalance that nobody even considers the negative consequences of turning a sleeper into an instakill machine

hmmmmmmmmmmmmm ok one second I'll change the machine so you have to put people into it akin to a disposal chute or a gibber, not just closing the lid but draggin' them on

**Comment by bobbahbrown** [comment]
ok @anconfuzedrock @Time-Green there is now a delay to drag people into the harvester, and it turns out the exiting action was already in place so I didn't need to touch that

**Comment by ATH1909** [comment]
> You can instantly sentence soneone to death by drag & dropping them on an emagged organ harvester, and no one would be any wiser.
> 
> The only counter would be jumping out very fast before it starts up, crowbarring it open or screaming over the radio and pray someone clones you.
> 
> But I kinda wanna see it happen though

Don't those grinders from disposals also basically instakill when emagged?

But yeah, I'm fine with there being a 3 second delay.

**Comment by ATH1909** [comment]
Say, in a future PR (I wouldn't want to cause the merging of this PR to be delayed), could we make the limbs and organs fly out into random nearby tiles instead of all (except for the torso) into the same tile?

**Comment by ATH1909** [comment]
Oh, I think I might have found a potential bug by looking at the code: if the organ harvester (even our current version of it) runs out of power mid-harvest, it will stop harvesting organs and will be rendered forever inoperable. As far as I can tell, if this happens while someone is inside of the harvester, they'll be trapped in there until something destroys the harvester (and they won't be able to resist out because "harvesting" would still be set to "TRUE".

This should be fixable by adding the bolded line of code:
```
/obj/machinery/harvester/proc/harvest()
	update_icon()
	if(!harvesting || state_open || !powered(EQUIP) || !occupant || !iscarbon(occupant))
		**end_harvesting()**
		return
	playsound(src, 'sound/machines/juicer.ogg', 20, 1)
	var/mob/living/carbon/C = occupant
	if(!LAZYLEN(operation_order)) //The list is empty, so we're done here
		end_harvesting()
		return
	var/turf/target
	for(var/adir in list(EAST,NORTH,SOUTH,WEST))
		var/turf/T = get_step(src,adir)
		if(!T)
			continue
		if(istype(T, /turf/closed))
			continue
		target = T
		break
	if(!target)
		target = get_turf(src)
	for(var/obj/item/bodypart/BP in operation_order) //first we do non-essential limbs
		BP.drop_limb()
		C.emote("scream")
		if(BP.body_zone != "chest")
			BP.forceMove(target)    //Move the limbs right next to it, except chest, that's a weird one
			BP.drop_organs()
		else
			for(var/obj/item/organ/O in BP.dismember())
				O.forceMove(target) //Some organs, like chest ones, are different so we need to manually move them
		operation_order.Remove(BP)
		break
	use_power(5000)
	addtimer(CALLBACK(src, .proc/harvest), interval)
```

**Comment by Time-Green** [comment]
@ATH1909 you can always crowbar them open

**Comment by bobbahbrown** [comment]
> Oh, I think I might have found a potential bug by looking at the code: if the organ harvester (even our current version of it) runs out of power mid-harvest, it will stop harvesting organs and will be rendered forever inoperable. As far as I can tell, if this happens while someone is inside of the harvester, they'll be trapped in there until something destroys the harvester, teleports them out of it, etc. (and they won't be able to resist out of the depowered (or depowered and then repowered) harvester because the "harvesting" variable would still be set to "TRUE").
> 
> This should be fixable by adding indicated line of code:
> 
> ```
> /obj/machinery/harvester/proc/harvest()
> 	update_icon()
> 	if(!harvesting || state_open || !powered(EQUIP) || !occupant || !iscarbon(occupant))
> 		end_harvesting()  //THIS IS THE ADDED LINE OF CODE
> 		return
> 	playsound(src, 'sound/machines/juicer.ogg', 20, 1)
> 	var/mob/living/carbon/C = occupant
> 	if(!LAZYLEN(operation_order)) //The list is empty, so we're done here
> 		end_harvesting()
> 		return
> 	var/turf/target
> 	for(var/adir in list(EAST,NORTH,SOUTH,WEST))
> 		var/turf/T = get_step(src,adir)
> 		if(!T)
> 			continue
> 		if(istype(T, /turf/closed))
> 			continue
> 		target = T
> 		break
> 	if(!target)
> 		target = get_turf(src)
> 	for(var/obj/item/bodypart/BP in operation_order) //first we do non-essential limbs
> 		BP.drop_limb()
> 		C.emote("scream")
> 		if(BP.body_zone != "chest")
> 			BP.forceMove(target)    //Move the limbs right next to it, except chest, that's a weird one
> 			BP.drop_organs()
> 		else
> 			for(var/obj/item/organ/O in BP.dismember())
> 				O.forceMove(target) //Some organs, like chest ones, are different so we need to manually move them
> 		operation_order.Remove(BP)
> 		break
> 	use_power(5000)
> 	addtimer(CALLBACK(src, .proc/harvest), interval)
> ```

I'll take a smell into it in a moment, I'm a little busy tonight but I should be able to figure it in

File: code/game/machinery/harvester.dm [hunk 1]
```diff
@@ -45,17 +45,34 @@
...
+			return
+		close_machine(target)
+		if (user == target)
+			user.visible_message("[user] climbs into [src].", "<span class='notice'>You climb into [src].</span>")
+		else
+			target.visible_message("<span class='danger'>[user] has placed [target] in [src].</span>", "<span class='userdanger'>[user] has placed [target] in [src].</span>")
+			log_combat(user, target, "forced", addition="into [src]")
+
+/obj/machinery/harvester/AltClick(mob/user)
+	if (!state_open && user != occupant)
```
Comment (AnturK) [inline code/game/machinery/harvester.dm]: No sanity for ghosts etc.

File: code/game/machinery/harvester.dm [hunk 2]
```diff
@@ -84,6 +101,10 @@
 		return
 	var/mob/living/carbon/C = occupant
 	operation_order = reverseList(C.bodyparts)   //Chest and head are first in bodyparts, so we invert it to make them suffer more
+	if(operation_order.len >= 2 && istype(operation_order[operation_order.len - 1], /obj/item/bodypart/head)) // Swap head and torso, so the head is processed last
```
Comment (AnturK) [inline code/game/machinery/harvester.dm]: There's a lot of assumptions in this code i don't like. Just find the head and chest then insert them at the end of the ordered list.

**Review by AnturK — CHANGES_REQUESTED** [review]

**Comment by ATH1909** [comment]
@bobbahbrown Why was this closed? Will it ever be reopened?

**Comment by bobbahbrown** [comment]
> @bobbahbrown Why was this closed? Will it ever be reopened?

closed for the time being as i didn't have time to make the changes, may reopen soon

**Comment by ATH1909** [comment]
@bobbahbrown Oof.

**Comment by TheBonded** [comment]
> > @bobbahbrown Why was this closed? Will it ever be reopened?
> 
> closed for the time being as i didn't have time to make the changes, may reopen soon

how soon is soon?

**Pull request closed without merging**